component/common/loki/http/config.go

```diff
@@ -0,0 +1,10 @@
+package http
+
+import (
+	"github.com/weaveworks/common/server"
+)
+
+// Config is a wrapper around server.Config.
+type Config struct {
```

Comment:
**incomplete**: the idea is for this `Config` object to expose a river block that is a subset of [this config](https://github.com/weaveworks/common/blob/master/server/server.go#L62). That way we can easily map it, extend it, and re-use the same kind of config block for all network loki sources. Let me know if that sounds good @tpaschalis ! 

Also, if yes, how easily can we break the river API from one version to another for things such as [loki.source.gcp](https://github.com/grafana/agent/blob/main/component/loki/source/gcplog/internal/gcplogtarget/types.go#L23)? Asking because I don't know which kind of api stability we have right now in flow

Reply:
I'd recommend that we go a slightly different way to be aligned with other Flow components work.

This struct can contain just the fields that Promtail's [server](https://grafana.com/docs/loki/next/clients/promtail/configuration/#server) block can already configure, but with River tags attached to them.

Then we can have an `UnmarshalRiver` method so that we apply defaults, performs validations etc implicitly during unmarshaling and finally a `Convert()` method so that it gets turned into a `github.com/weaveworks/common/server.Config` that the caller packages can use instead of building the config by hand. If this sounds complicated, [here's an example](https://github.com/grafana/agent/blob/6fc177a56a2f71c8980fac266bb23d78b6572f75/component/common/config/types.go#L19-L140) of doing that with our shared `HTTPClientConfig` struct. 

Then, packages like gcplog could embed this config in their Arguments, call Convert() to get the config and then call to NewTargetServer.

We could even go a step further and have the `NewTargetServer` receive the River struct itself and call Convert() inside of it. That means that caller packages don't have to construct the config themselves like we do [here](https://github.com/grafana/agent/pull/3581/files#diff-fb88003d94f0a346752ac1519c4b80e685c763f8e716a83a935934a9117143aeR43-R49) which has leaves some room for error.

Reply:
As for the component Arguments, I don't think there's too much hassle around changing the config, especially if it's for a good reason such as here. Flow itself as a whole is stable, but I believe that provided we document and communicate the change we should be okay.

Reply:
Completely agree with the suggested approach. I went ahead and first renamed `Config` to `ServerConfig`, which exposes in River a some configuration blocks for tuning the underlying server. I haven't added all settings since it's a huge amount, and rather just used the ones we found the most usable for a scalable server (for http listen and port, and for grpc some more that control connections lifecycle).

Also, routed this new settings through heroku and gcp components. Let me know if we should keep backwards with this both, and we can discuss how, I'm assuming since we can break it between one release and the other.